pennylane/data/__init__.py

```diff
@@ -24,7 +23,177 @@
+            doc="Eigenvalues and eigenvectors of the hamiltonain")
+
+This metadata can then be accessed using the :meth:`Dataset.attr_info` mapping:
+
+    >>> dataset.attr_info["hamiltonian"]["doc"]
+    'The hamiltonian of the system'
+
+
+Declarative API
+~~~~~~~~~~~~~~~
+
+When creating datasets to model a physical system, it is common to collect the same data for
+a system under different conditions or assumptions. For example, a collection of datasets describing
+a quantum oscillator, which contains the first 1000 energy levels for different masses and force constants.
+
+The datasets declarative API allows us to create subclasses of ``Dataset`` that define the required attributes,
+or 'fields', and their associated type and documentation:
+
+.. code-block
+
+    class QuantumOscillator(qml.data.Dataset, data_name="quantum_oscillator", params=["mass", "force_constant"]):
+        \"""Dataset describing a quantum oscillator.\"""
+
+        mass: float = qml.data.field(doc = "The mass of the particle")
+        force_constant: float = qml.data.field(doc = "The force constant of the oscillator")
+        hamiltonian: qml.Hamiltonian = qml.data.field(doc = "The hamiltonian of the particle")
+        energy_levels: np.ndarray = qml.data.field(doc = "The first 1000 energy levels of the system")
+
+The ``data_name`` keyword specifies a category or descriptive name for the dataset type, and the ``params``
+keyword to the class is used to specify fields that function as parameters, i.e they determine the behaviour
+of the system.
+
+When a ``QuantumOscillator`` dataset is created, its attributes will have the documentation from the field
+definition:
+
+    >>> dataset = QuantumOscillator(mass=1, force_constant=0.5, hamiltonian=..., energy_levels=...)
+    >>> dataset.attr_info["mass"]["doc"]
+    'The mass of the particle'
+
 """
 
-from .dataset import Dataset
-from .data_manager import load, load_interactive, list_datasets, list_attributes
+from .attributes import (
+    DatasetArray,
+    DatasetDict,
+    DatasetJSON,
+    DatasetList,
+    DatasetMolecule,
+    DatasetNone,
+    DatasetOperator,
+    DatasetScalar,
+    DatasetSparseArray,
+    DatasetString,
+)
+from .base import DatasetNotWriteableError
+from .base.attribute import AttributeInfo, DatasetAttribute, attribute
+from .base.dataset import Dataset, field
+from .data_manager import (
+    DEFAULT,
+    FULL,
+    list_attributes,
+    list_datasets,
+    load,
+    load_interactive,
+)
+
+__all__ = (
+    "AttributeInfo",
+    "attribute",
+    "field",
+    "Dataset",
+    "DatasetAttribute",
+    "DatasetNotWriteableError",
+    "DatasetArray",
+    "DatasetScalar",
+    "DatasetString",
+    "DatasetList",
+    "DatasetDict",
+    "DatasetOperator",
+    "DatasetNone",
+    "DatasetMolecule",
+    "DatasetSparseArray",
+    "DatasetJSON",
+    "load",
+    "load_interactive",
+    "list_attributes",
+    "list_datasets",
+    "DEFAULT",
```

Comment:
Do we already have data for supporting `DEFAULT` for the s3 backend? 🤔

Reply:
yes, you can try it out now:
```
qchem.data.load('qchem', molname='O2')
```
will return the default dataset for `O2`